File: leetcode solutions/Python/swap nodes in pairs/swapNodesInPairs.py
class ListNode(object):
    def __init__(self, x):
        self.val
        self.next = None


class Solution(object):
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head or not head.next:
            return head

        dummy = ListNode(0)
        dummy.next = head
        curr = dummy

        while curr.next and curr.next.next:
            nextOne = curr.next
            nextTwo = curr.next.next
            nextThree = curr.next.next.next

            curr.next = nextTwo
            nextTwo.next = nextOne
            nextOne.next = nextThree
            curr = nextOne

        return dummy.next

File: leetcode solutions/C++/swap nodes in pairs/swapNodesInPairs.cpp
struct ListNode {
    int val;
    ListNode *next;
    ListNode (int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* swapPairs(ListNode* head) {

        if ((head == NULL) || (head->next == NULL)) {
            return head;
        }

        ListNode *dummy = new ListNode(0);
        dummy->next = head;
        ListNode *curr = dummy;
        ListNode *nextOne = NULL;
        ListNode *nextTwo = NULL;
        ListNode *nextThree = NULL;

        while (curr->next != NULL && curr->next->next != NULL) {
            nextOne = curr->next;
            nextTwo = curr->next->next;
            nextThree = curr->next->next->next;

            curr->next = nextTwo;
            nextTwo->next = nextOne;
            nextOne->next = nextThree;

            curr = nextOne;
        }
        return dummy->next;
    }
};